### PR TITLE
[doc] Update doc with steps for custom metric collection

### PR DESCRIPTION
**Pull request by varshaprasad96**
**Description of the change:**
This PR updates the user documentation with the procedure
to collect operator specific metrics using the metrics package
provided by SDK.

**Motivation for the change:**
Currently, many of the operator authors/owners are unaware
of the procedure to leverage `metrics package` provided by 
SDK to collect operator specific metrics. This PR would update
the documentation with the required steps to enable operator
specific custom metrics collection using the functions provided
by the SDK.

### CODE REVIEW

File: doc/user/metrics/README.md [hunk 1]
```diff
@@ -86,3 +86,124 @@ By default operator will expose info metrics based on the number of the current
...
+
+        // Pass metrics address to controller-runtime manager
+        mgr, err := manager.New(cfg, manager.Options{
+            Namespace:          namespace,
+        })
+
+        ...
+
+        // Add to the below struct any other metrics ports you want to expose.
+	    servicePorts := []v1.ServicePort{
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]: I think it need to be indented.

File: doc/user/metrics/README.md [hunk 2]
```diff
@@ -86,3 +86,124 @@ By default operator will expose info metrics based on the number of the current
...
+  - create
+
+```
+
+
+#### Custom metric collection
+
+Implement the logic to collect specific custom metrics inside the operator code base. For example, the following is a snippet from `Marketplace` operator which implements custom metric collection:
+
+```
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
```Go
```

File: doc/user/metrics/README.md [hunk 3]
```diff
@@ -86,3 +86,124 @@ By default operator will expose info metrics based on the number of the current
...
+	)
+
+    metricsList = []prometheus.Collector{
+		errorMetric,
+	}
+)
+```
+Register the metrics created above with Prometheus:
+
+```
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
```Go
```

File: doc/user/metrics/README.md [hunk 4]
```diff
@@ -86,3 +86,124 @@ By default operator will expose info metrics based on the number of the current
...
+
+        // Add to the below struct any other metrics ports you want to expose.
+	    servicePorts := []v1.ServicePort{
+		    {Port: metricsPort, Name: metrics.OperatorPortName, Protocol: v1.ProtocolTCP, TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: metricsPort}},
+	    }
+
+        // Create Service object to expose the metrics port.
+        _, err = metrics.CreateMetricsService(context.TODO(), cfg, servicePorts)
+        if err != nil {
+            // handle error
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
            // TODO (user): handle error
```

File: doc/user/metrics/README.md [hunk 5]
```diff
@@ -86,3 +86,124 @@ By default operator will expose info metrics based on the number of the current
...
+4. Expose metrics: Utilize the functions provided by SDK to expose metrics and start the metrics server.
+
+#### Update the Operator
+
+Update the operator and provide sufficient permissions to allow creation of Prometheus resources. The required permissions are:
+1. `Service Monitor` : `get` and `create`
+2. `roles` and `rolebindings` : `get`, `create`, `delete`, `update`, `list`
+
+The below is a snippet from roles.yaml:
+```
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
```YAML
```

File: doc/user/metrics/README.md [hunk 6]
```diff
@@ -86,3 +86,124 @@ By default operator will expose info metrics based on the number of the current
...
+
+
+#### Custom metric collection
+
+Implement the logic to collect specific custom metrics inside the operator code base. For example, the following is a snippet from `Marketplace` operator which implements custom metric collection:
+
+```
+// Metric variables and metric list that are used to report error. 
+var (
+	errorMetric = prometheus.NewCounterVec(
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]: It shows very nice, but IMHO the indentation should equal it is done in the other docs. See here: https://github.com/operator-framework/operator-sdk/blob/master/doc/user/unit-testing.md

**Comment by varshaprasad96** [comment]
@camilamacedo86 - I have updated the PR and addressed your comments. Please have a look at it.

File: doc/user/metrics/README.md (hunk 6 above)
Comment (camilamacedo86) [reply doc/user/metrics/README.md]: Solved

File: doc/user/metrics/README.md [hunk 7]
```diff
@@ -39,9 +39,9 @@ By default, the metrics are served on `0.0.0.0:8383/metrics`. To modify the port
         ...
 
         // Add to the below struct any other metrics ports you want to expose.
-	    servicePorts := []v1.ServicePort{
-		    {Port: metricsPort, Name: metrics.OperatorPortName, Protocol: v1.ProtocolTCP, TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: metricsPort}},
-	    }
+	 servicePorts := []v1.ServicePort{
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
	servicePorts := []v1.ServicePort{
```

Comment (camilamacedo86) [reply doc/user/metrics/README.md]: Small nit in the indentation here.

File: doc/user/metrics/README.md [hunk 8]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+	    }
+
+        // Create Service object to expose the metrics port.
+        _, err = metrics.CreateMetricsService(context.TODO(), cfg, servicePorts)
+        if err != nil {
+            // TODO (user): handle error
+        }
+
+        // Start the metrics server
+        http.Handle(MetricsPath, promhttp.Handler())
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]: from where came this MetricsPath? It is not working.

File: doc/user/metrics/README.md [hunk 9]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+
+        // Add to the below struct any other metrics ports you want to expose.
+	    servicePorts := []v1.ServicePort{
+		    {Port: metricsPort, Name: metrics.OperatorPortName, Protocol: v1.ProtocolTCP, TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: metricsPort}},
+	    }
+
+        // Create Service object to expose the metrics port.
+        _, err = metrics.CreateMetricsService(context.TODO(), cfg, servicePorts)
+        if err != nil {
+            // TODO (user): handle error
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
           log.Error(err, "")

```

File: doc/user/metrics/README.md [hunk 10]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+  - update
+  - list
+  - watch
+- apiGroups:
+  - monitoring.coreos.com
+  resources:
+  - servicemonitors
+  verbs:
+  - get
+  - create
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]: The following part is already present in the default roles added in the Role.yaml, so should it be really required to be added here? Should not better just add what is required to do in order to make it work?  

```
- apiGroups:
  - monitoring.coreos.com
  resources:
  - servicemonitors
  verbs:
  - get
  - create
```
IMHO: we could remove this part.

**Comment by camilamacedo86** [comment]
Hi @@varshaprasad96, 

Really thank you for your contribution. However, could you please check the review comments?

File: doc/user/metrics/README.md [hunk 11]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
   - update
 ...
 ```
-
 ### Custom resource specific metrics
 
 By default operator will expose info metrics based on the number of the current instances of an operator's custom resources in the cluster. It leverages [kube-state-metrics][ksm] as a library to generate those metrics. Metrics initialization lives in the `cmd/manager/main.go` file of the operator in the `serveCRMetrics` function. Its arguments are a custom resource's group, version, and kind to generate the metrics. The metrics are served on `0.0.0.0:8686/metrics` by default. To modify the exposed metrics port number, change the `operatorMetricsPort` variable at the top of the `cmd/manager/main.go` file in the generated operator.
 
+### Custom Metrics using Operator-SDK
+To implement custom metric collection using the `metrics package` provided by SDK, follow the steps outlined below:
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
Following the steps to implement custom metric collection using the `metrics package` provided by SDK.
```

File: doc/user/metrics/README.md [hunk 12]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+1. Update the Operator: Update the operator to provide sufficient permisisons so that the required resources for prometheus can be created.
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]: What permission is required that it not added by default?

File: doc/user/metrics/README.md [hunk 13]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+2. Custom metric collection: Implement the logic for custom metric variables defined for your operator and register the specific metrics with prometheus. 
+3. Expose metrics: Utilize the functions provided by SDK to expose metrics and start the metrics server.
+
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]: IMHO has no need for the above steps since they will be described bellow

File: doc/user/metrics/README.md [hunk 14]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+#### Update the Operator
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
#### Add RBCA Permissions
```

File: doc/user/metrics/README.md [hunk 15]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+
+Update the operator and provide sufficient permissions to allow creation of Prometheus resources. The required permissions are:
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
Ensure that `role.yaml` file has the following permissions. 
```

File: doc/user/metrics/README.md [hunk 16]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+- apiGroups:
+  - monitoring.coreos.com
+  resources:
+  - servicemonitors
+  verbs:
+  - get
+  - create
+
+```
+#### Custom metric collection
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
#### Developing Customized Metrics 
```

File: doc/user/metrics/README.md [hunk 17]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+
+Implement the logic to collect specific custom metrics inside the operator code base. For example, the following is a snippet from `Marketplace` operator which implements custom metric collection:
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
Following an example of business logic implementation for customized metrics.  
```

Comment (camilamacedo86) [reply doc/user/metrics/README.md]: Could you please add an example based on the Memcached sample? 

Also, WDYT to add a NOTE with the link for the code source where has this impl in the `Marketplace` project as the following example?

 ```suggestion
...
**NOTE:** To check a further example see it implemented [here][] in the [Marketplace][marketplace] project  
```

File: doc/user/metrics/README.md [hunk 18]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+	for _, metric := range metricsList {
+		err := prometheus.Register(metric)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+```
+#### Expose metrics
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]:
```suggestion
#### Exposing customized metrics
```

File: doc/user/metrics/README.md [hunk 19]
```diff
@@ -74,15 +74,129 @@ In Kubernetes clusters where [OwnerReferencesPermissionEnforcement][ownerref-per
...
+The steps provided previously for creating a service and service monitor resource can be followed to expose custom metrics. Additionally, the metrics server has to be manually started instead of providing it as an argument to the controller-runtime manager.
+Usage:
+
```
Comment (camilamacedo86) [inline doc/user/metrics/README.md]: Does NOT the following code generated by default in the scalfod project? WDYT about adding an example of a customized metric above using the Memcached sample and them show here just what needs to be done to expose it?

**Comment by varshaprasad96** [comment]
I am extremely sorry for the delay. I will make the changes and update the PR.

**Comment by camilamacedo86** [comment]
Hi @varshaprasad96, 

Really thank for your contribution 🥇.
It is very important and helpful.

File: doc/user/metrics/README.md (hunk 12 above)
Comment (varshaprasad96) [reply doc/user/metrics/README.md]: I have removed this part, as the required permissions are added by default.

File: doc/user/metrics/README.md (hunk 17 above)
Comment (varshaprasad96) [reply doc/user/metrics/README.md]: Im not sure on whether the metrics collection is completely done on marketplace. I have provided a CounterVec as an example. Users have directly use this inside their operator. I'm not sure on what is required with specific to Memcached sample. Could you please elaborate more on that?

File: doc/user/metrics/README.md (hunk 19 above)
Comment (varshaprasad96) [reply doc/user/metrics/README.md]: Yes, the code is now generated by default in the scaffold project. But the server has to still be started again. I have provided the code just as an example to do so.

Comment (camilamacedo86) [reply doc/user/metrics/README.md]: It shows fine.

File: doc/user/metrics/README.md (hunk 17 above)
Comment (camilamacedo86) [reply doc/user/metrics/README.md]: But you added `following is a snippet from `Marketplace`, so would be great to have the link for this piece of code there and the `Marketplace`. Or, it should not be referenced.  is it make sense?

**Comment by camilamacedo86** [comment]
Hi @varshaprasad96,

It still missing some points to address in the order we are able to move forward and merge it. Could you please check them?

**Comment by camilamacedo86** [comment]
Hi @varshaprasad96, 

Currently, the project is looking for to contribute with k8s docs and address the users check them when the feature is based on its implementation. So, after a deeper analysis in this PR shows that these details could be better addressed in https://book.kubebuilder.io/reference/metrics.html#publishing-additional-metrics. See that its docs will redirect the users to check this document. 

Then, because of this, we are closing this PR. However, please feel free to re-open if you think that something should be done here instead of.

**Comment by varshaprasad96** [comment]
Hi @camilamacedo86,

Thanks for the update. I see that the process is being explained in the kubebuilder docs. 

Just curious to know - the doc states that the custom metrics could be registered by declaring the global variables and registering them using `MustRegister` function of Prometheus. Wouldn't this collide with the default metrics, as both would get registered in the same port unless specified explicitly?

Maybe I am missing something, but just got this doubt when I saw the doc ! :)

**Pull request closed without merging**